### PR TITLE
Adding order related operations to gui bindings

### PR DESCRIPTION
**Pull request by findolor**
## Motivation

See issue: #964

## Solution

- Added various operations for order related actions

## Checks

By submitting this for review, I'm confirming I've done the following:
- [x] made this PR as small as possible
- [x] unit-tested any new functionality
- [x] linked any relevant issues or PRs
- [ ] included screenshots (if this involves a front-end change)

### CODE REVIEW

File: crates/js_api/src/gui/order_operations.rs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            .outputs
+            .iter()
+            .filter(|output| !all_vaults.contains(*output))
+            .cloned()
+            .collect();
+        all_vaults.extend(filtered_outputs);
+        all_vaults
+    }
+
+    fn get_deposits_as_map(&self) -> HashMap<Address, String> {
```
Comment (hardyjosh) [inline crates/js_api/src/gui/order_operations.rs:66]: deposits should be U256

File: crates/js_api/src/gui/order_operations.rs [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        Ok(decimals)
+    }
+
+    async fn get_deposit_args(
+        &self,
+        rpc_url: &str,
+        deposits_map: &HashMap<Address, String>,
+        order_io: &OrderIO,
+    ) -> Result<DepositArgs, GuiError> {
+        let decimals = if let Some(decimals) = order_io.token.decimals {
```
Comment (hardyjosh) [inline crates/js_api/src/gui/order_operations.rs:92]: it's not really safe to use the decimals in the yaml - i think we should always check onchain

what would be good is to store all the token info (name/symbol/decimals) in memory, and allow it to be prefetched. then here we would either use the info already in memory, or fetch if we have to

File: crates/js_api/src/gui/order_operations.rs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+            allowance,
+        })
+    }
+
+    /// Check allowances for all inputs and outputs of the order
+    ///
+    /// Returns a vector of [`TokenAllowance`] objects
+    #[wasm_bindgen(js_name = "checkAllowances")]
+    pub async fn check_allowances(&self, owner: String) -> Result<JsValue, GuiError> {
+        let all_vaults = self.get_all_vaults();
```
Comment (hardyjosh) [inline crates/js_api/src/gui/order_operations.rs:139]: instead of checking allowances for all vaults, shouldn't we be checking allowances for all deposits?

we don't need to check allowance for input vault tokens (that's for tokens coming in, so you don't deposit into an input vault). we also don't need to check allowance for output vault tokens that aren't being deposited into.

File: crates/js_api/src/gui/order_operations.rs [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+    }
+
+    fn get_deposits_as_map(&self) -> HashMap<Address, String> {
+        self.get_deposits()
+            .into_iter()
+            .map(|d| (d.address, d.amount))
+            .collect()
+    }
+
+    async fn get_token_decimals(&self, rpc_url: &str, token: Address) -> Result<u8, GuiError> {
```
[the remark this answers is not in the document]
Comment (findolor) [reply crates/js_api/src/gui/order_operations.rs:73]: yeah this makes a lot of sense since we can just use that in other places in the application. will add that

File: crates/js_api/src/gui/order_operations.rs (hunk 3 above)
Comment (findolor) [reply crates/js_api/src/gui/order_operations.rs:139]: ok i wasn't sure before implementing this but i get the picture now, will do it based on this

File: crates/js_api/src/gui/order_operations.rs [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+
+    async fn get_deposit_calldatas(
+        &self,
+        vaults_and_deposits: &Vec<(OrderIO, U256)>,
+    ) -> Result<Vec<Vec<u8>>, GuiError> {
+        let mut results = Vec::new();
+        for (order_io, amount) in vaults_and_deposits.iter() {
+            let deposit_args = DepositArgs {
+                token: order_io.token.address,
+                vault_id: order_io.vault_id.unwrap_or(rand::random()),
```
Comment (hardyjosh) [inline crates/js_api/src/gui/order_operations.rs:176]: why is this needed?

File: crates/js_api/src/gui/order_operations.rs [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+            let allowance = self
+                .check_allowance(&orderbook, &deposit_args, &owner)
+                .await?;
+            results.push(allowance);
+        }
+
+        Ok(serde_wasm_bindgen::to_value(&results)?)
+    }
+
+    /// Generate approval calldatas for all inputs and outputs of the order
```
Comment (hardyjosh) [inline crates/js_api/src/gui/order_operations.rs:129]: still says inputs?

**Review by hardyjosh — APPROVED** [review]
